### PR TITLE
[JENKINS-33495 and JENKINS-33496] Separate scrollspy widget code out from tab widget code

### PR DESCRIPTION
**Pull request by tfennelly**
See [JENKINS-33495](https://issues.jenkins-ci.org/browse/JENKINS-33495) and [JENKINS-33496](https://issues.jenkins-ci.org/browse/JENKINS-33496).
- [x] Revert original code committed from #2106. All of the scrollspy js and css/less committed with that PR was in the wrong place i.e. roughed in on top of the tab code. Easier to revert back out and then add back in the right place versus trying to tease them appart by hand.
- [x] Create scrollspy specific widget components i.e. js, css/less.
- [x] Add scrollspy tests.

### CODE REVIEW

**Comment by tfennelly** [comment]
@jenkinsci/code-reviewers and @reviewbybees

**Comment by rtyler** [comment]
When I run this locally there's a "Bindings" tab which doesn't correspond to anything in the UI that I can see. It also scrolls "up", always up

**Comment by rtyler** [comment]
screencast

**Comment by tfennelly** [comment]
@rtyler oh funky. So what plugin contributes that "Binding" section and is it visible if you remove the tabs (press the "remove tabs" button)?

**Comment by tfennelly** [comment]
@rtyler and does that strange auto-scroll only happen when you mouse over the "Bindings" tab?

**Comment by rtyler** [comment]
@tfennelly I'm clicking on the "Bindings" tab in the Gif

**Comment by daniel-beck** [comment]
It's the Credentials Binding plugin, and I usually see it show up as a section.

**Comment by rtyler** [comment]
> 17:12 <@danielbeck> rtyler Could you restart? This is Crednetials Binding Pluign, maybe something broken there?

After a restart I see the same behavior

**Comment by rtyler** [comment]
I should note, that I'm using a pristine `JENKINS_HOME` in my gif with only one job

**Comment by tfennelly** [comment]
@rtyler do you see the section when tabs are removed?

**Comment by rtyler** [comment]
@tfennelly Negative

**Comment by tfennelly** [comment]
@rtyler ok and if you view the HTML source, can you se the rows that correspond to that plugin? Looks like there's something there in the source, between "Build Environment" and "Build".

**Comment by rtyler** [comment]
@tfennelly I'm sure you can reproduce this locally, driving my browser through pull request comments would probably take too long :)

1. Take [the build from your pull request](https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/4356/artifact/war/target/jenkins.war)
1. Create a fresh directory, e.g. `/tmp/jenkins`
1. Run: `JENKINS_HOME=/tmp/jenkins java -jar jenkins.war`
1. Go through Getting Started dialogue and select "Install suggested plugins"
1. Create a new freestyle job

**Comment by tfennelly** [comment]
@rtyler @daniel-beck after turning off tabs I see that this Binding stuff is hidden by default for some reason. 

I was able to manual remove the `display: none` style and make it appear, but it does seem like something else is wrong (maybe with that plugin?).

I'll see can I find a way of not displaying a tab. Could probably just check the `display` on the section header row.

**Comment by kzantow** [comment]
@tfennelly $(<thing>).is(':visible')

**Comment by tfennelly** [comment]
@kzantow thanks man ... will try that. Fairly sure I tried using that before and had no success, but maybe I was doing something wrong.

**Comment by tfennelly** [comment]
@rtyler @daniel-beck so tabs are no longer displayed for "hidden" sections. PR build in progress (also includes hiding of the finder for scrollspy).

@kzantow thanks for saving me the bother of looking for `$().is(':visible')`. Worked a charm :)

**Comment by gusreiber** [comment]
It would be better if the "Advanced Project Options" was not its own tab, but rather was lumped under "General". 
advanced-under-general

**Comment by gusreiber** [comment]
Ideally we would have some way to manage tab size. "Source Code Management"  is
 ridiculously large. "SCM" or just "Source Code" would be a lot better. Not sure where the magic would come from to do that truncation, but it would be worth doing. Tabs need some sort of maximum size. "Adanced...blah... blah..." is also silly long, but I think we can safely just nuke him.
screen shot 2016-03-16 at 12 41 12 pm

**Comment by kzantow** [comment]
@gusreiber you can use CSS for this pretty easily, as I suspect you know: `max-width: X; white-space: nowrap; text-overflow: ellipsis;`

**Comment by gusreiber** [comment]
@kzantow @tfennelly , yeah, that sorta sucks, though. For a number of reasons, not least of which is that the ellipsis take up a decent amount of width. Would be much better to do some sort of either custom handling, or algorithm (if char count is over blah, plus word has spaces...) gets tough and messy pretty fast, but not impossible. Could also do forced overlapping with complete text on hover, ala, my chrome web browser.  ...doing nothing is always an option as well, but not my favorite.

**Comment by daniel-beck** [comment]
@gusreiber 

> It would be better if the "Advanced Project Options" was not its own tab, but rather was lumped under "General". 

As it is a section, and every section gets its title, I asked for this to be restored. In tabs, it was weird, here it makes sense. However, that was before you decided to limit the width, so it wasn't a problem then.

**Comment by gusreiber** [comment]
I think we can drop the 'remove tabs' now that they can be scrolled.
screen shot 2016-03-16 at 1 01 03 pm

**Comment by gusreiber** [comment]
@daniel-beck  ....but that is crazy. So let us not do that.

**Comment by daniel-beck** [comment]
For some reason it is possible to scroll into an area where no section is considered active. This is an empty job config form in the PR build after installing all plugins but Slack Plugin. Seems to be dependent on browser window size or zoom level.

screen shot 2016-03-16 at 21 00 14

**Comment by daniel-beck** [comment]
It's not clear to me why we're still showing the *unstyled* config form while loading. Why not make the overlay opaque?

screen shot

**Comment by gusreiber** [comment]
@daniel-beck  ...by not, I mean, not-not-not..... let's not give precious space to the category that is specifically designed to remove clutter from the screen. That is totally silly. Obviously so.

**Comment by daniel-beck** [comment]
Sometimes the config form still does not get its styling after loading is finished. Seems to happen when page loads take longer, so maybe install a large number of plugins to make it happen?

screen shot

**Comment by daniel-beck** [comment]
@gusreiber Given the number and magnitude of real issues this still has, it looks to me a lot like you're wasting your time on basically nothing.

**Comment by gusreiber** [comment]
@daniel-beck I think the reason to not make it opaque is that it is possible for JS, especially JS in Jenkins to break mysteriously. With it semi-opaque, the user knows the form is there and if they are clever, can work around it. Same logic I took to the item choice page, better to not kill the underlying DOM. ....that having been said, we chose to kill the underlying DOM there on the logic, 'make it work or bust' ...which is totally good logic. ....but that is the reasoning we could reject.

**Comment by daniel-beck** [comment]
@gusreiber I've already made my arguments regarding that on https://github.com/jenkinsci/jenkins/pull/1885

**Comment by gusreiber** [comment]
@daniel-beck ...fair enough. But the horizontal size of the tabs is essential to its usability. I am not prioritizing, I am just listing. Your argument in #1885 was equally wrong then as it is now.

**Comment by tfennelly** [comment]
> Sometimes the config form still does not get its styling after loading is finished. Seems to happen when page loads take longer, so maybe install a large number of plugins to make it happen?

I eventually managed to reproduce this on Safari (only). It's something to do with prototypejs' `Behaviour.specify` where the callback never gets called. Maybe there's a timeout on it.  Looking through the source.

**Comment by kzantow** [comment]
@tfennelly FYI - I saw this happen once, where it took 30-60 seconds before the tabs showed up, maybe there's a download blocking it from calling a page complete or whatever the Prototype equivalent is?

**Comment by tfennelly** [comment]
@kzantow no idea yet. It's def some issue with prototype, but not sure what it is yet.

**Comment by gusreiber** [comment]
@daniel-beck @tfennelly ...so Daniel and I punched each other in the head for a bit and now I feel a lot better.... I think Daniel's objection is that the tabs should always follow the same rules. But he is OK with yanking that "Advanced Project Settings" header form the jelly page. I am thinking this is all the sort of light clean-up work I can do at the end, assuming no one disagrees. I am worried about the amount of horizontal space the tabs might need, thus my focus here on that concern.

**Comment by daniel-beck** [comment]
I noticed the browser window cannot be resized to the config form's actual width without allowing *a lot* of horizontal scrolling.

**Comment by tfennelly** [comment]
@daniel-beck I think all of the main issues are resolved now. I'm on vacation atm, but will check in later. Feel free to merge if you're "happy" with it.

**Comment by daniel-beck** [comment]
With the current PR build (4377) I am unable to reproduce the "style does not get applied" issue :smile: 

There was still some weirdness when toggling visibility of the "Bindings" section, but I see there's a new commit that appears related.

**Comment by daniel-beck** [comment]
@gusreiber Long lines in formatted text fields still break the layout as determined by @jglick in a previous PR.

screen shot

**Comment by tfennelly** [comment]
@daniel-beck unless there's some major issue here it'd be nice to get this merged.

**Comment by tfennelly** [comment]
> There was still some weirdness when toggling visibility of the "Bindings" section

I'm afraid that's more-or-less impossible to make "unweird" since those sections are initially visible and then become hidden.

**Comment by daniel-beck** [comment]
> I'm afraid that's more-or-less impossible to make "unweird" since that section is initially visible and then becomes hidden.

Depends on whether the related option in the previous section starts out checked or unchecked.

**Comment by tfennelly** [comment]
> Depends on whether the related option in the previous section starts out checked or unchecked.

@daniel-beck ok ... I'm not sure what you mean by that.

**Comment by gusreiber** [comment]
@daniel-beck @kzantow Keith is grabbing the heavier lifting here, I am going to grab the loader here and the transparent goofiness. The tab wrapping possibility is bad news, but we are not sure if there is a cheap enough fix, so we are thinking a unification of Add-item and Config-item PR might be worth doing. Some of these usability issues would be re-addressed there.

**Comment by daniel-beck** [comment]
@gusreiber Please keep independent changes in independent PRs.

**Comment by gusreiber** [comment]
@daniel-beck ...well, that is sort of my point. Some of the feedback here is about the scrolling-tabs behavior, which is behavior 2 different PRs share. Also, some aspects commented on here could be address via CSS, which 2 PRs definately share. ...so my ask here is wouldn't it make some sense to merge this PR as soon as there are no show-stoppers and then look at additional PRs to smooth shared issues?

**Comment by daniel-beck** [comment]
Observed in build 4393: The tab bar does not move when the page is scrolled programmatically from the top position until the scroll is finished.

**Steps to reproduce**:
1. Go to config page
2. Click any of the titles (except General)

**Expected**: Page scrolls down, tab bar :detaches" from its regular location and gets fixed position near the top and stays there.
**Actual**: Page scrolls down, tab bar disappears until scroll is finished, then it pops up again.

I'm pretty sure this is a recent regression in this PR.

**Comment by tfennelly** [comment]
@daniel-beck yes that is different behaviour but has been there since the earliest builds on this PR, trying to fix the jumpiness that you highlighted :)

Anyway ... you forced me to rethink and find a better way of doing it. I just pushed a fix for that.

TBH ... I think we are nit-picking here now at this stage. Can we please get this merged. Can @kzantow and @gusreiber give it bees or not so we can merge it. I'm tempted to just follow KKs lead at this stage and just merge.

@gusreiber you were also confusing me there when talking about other fixes and features that could be added to this stuff (config pages, scrollspy etc). It's standard practise to not talk about other things on a PR and to just keep the conversation fixed on the changes in the PR.

**Comment by tfennelly** [comment]
I'm going to look back here in a few hours and merge this PR if it's not already merged.

**Comment by daniel-beck** [comment]
@tfennelly I don't see a single code review related comment or :bee: so far. Tyler and I tested it, but I doubt either of us can actually *review* the changes.

I asked Gus last night to 1. document open issues in Jira so they don't get lost, and 2. find people to actually code review this and approve. AFAICT, neither has happened so far.

File: core/src/main/resources/lib/layout/layout.jelly [hunk 1]
```diff
@@ -99,13 +99,7 @@ ${h.initPageVariables(context)}
...
     <link rel="shortcut icon" href="${resURL}/favicon.ico" type="image/vnd.microsoft.icon" />
     <link rel="mask-icon" href="${rootURL}/images/mask-icon.svg" color="black" />
-    
-    <!-- should the sidebar be hidden on load? -->
-    <style>
-      body.j-hide-left #side-panel{width:0; padding:0; overflow:hidden;}
-      body.j-hide-left #main-panel{margin:0 auto; max-width:75em;}
-    </style>
-    
+
```
Comment (recena) [inline core/src/main/resources/lib/layout/layout.jelly]: @tfennelly Thanks for removing that.

File: war/src/main/webapp/css/style.css [hunk 2]
```diff
@@ -1932,4 +1931,14 @@ table#legend-table td {
...
+}
+
+body.main-panel-only #side-panel {
+    width: 0 !important;
+    padding: 0 !important;
+    overflow: hidden !important;
+}
+body.main-panel-only #main-panel {
+    margin: 0 auto !important;
+    max-width: 75em !important;
```
Comment (recena) [inline war/src/main/webapp/css/style.css]: How does it work with? https://github.com/jenkinsci/jenkins/pull/2029

We should merge https://github.com/jenkinsci/jenkins/pull/2029 before this.

/cc @daniel-beck

File: core/src/main/resources/hudson/model/Job/configure.jelly [hunk 3]
```diff
@@ -27,14 +27,14 @@ THE SOFTWARE.
...
     <f:breadcrumb-config-outline />
     <l:main-panel>
-      <l:js src="jsbundles/config-tabbar.js" />
-      <l:css src="jsbundles/jenkins-widgets.css" />
-      <div class="behavior-loading">${%loading}</div>
-      <f:form method="post" action="configSubmit" name="config" tableClass="job-config tabbed">
+      <l:js src="jsbundles/config-scrollspy.js" />
+      <l:css src="jsbundles/config-scrollspy.css" />
+
+      <div class="behavior-loading">${%LOADING}</div>
```
Comment (daniel-beck) [inline core/src/main/resources/hudson/model/Job/configure.jelly]: I expect this to break all existing translations for no real reason. Could just be changed in the corresponding `.properties` file.

**Comment by gusreiber** [comment]
@daniel-beck @tfennelly Daniel is correct.  I need to work through a review in some depth. Will be doing so here today.

**Comment by daniel-beck** [comment]
I've been told @kzantow is currently reviewing this as well. If he amends this PR, I'd appreciate reviews of that by @gusreiber. This way we at least have some review coverage here, and I'm more comfortable merging this.

**Comment by kzantow** [comment]
I've found a blocking issue: something is causing the draggable behavior to be unusable. This is a regression from what's in the current 2.0 brach. I'm currently investigating this, but I can't give the ok to merge until it's sorted out.

**Comment by gusreiber** [comment]
@kzantow @tfennelly @daniel-beck I see the issue @kzantow is point to. I wrestled with it in my rough version. I will have a fix here shortly.

**Comment by kzantow** [comment]
Ok, I have a fix for this but I can't seem to actually push to this repo or make a PR against it... https://github.com/kzantow/jenkins/commit/05a660c8529239bb7daad9d945a06308230e43eb

**Comment by kzantow** [comment]
Ok, I've reviewed this, and agree with Tom we should get it merged in, this PR is too big to add anything further to be able to meaningfully review, and as is I know the pending fix for the only blocking issue and can follow it up immediately.
:bee: :+1:

**Comment by kzantow** [comment]
Note: I've created https://issues.jenkins-ci.org/browse/JENKINS-33659 to address the width issues mentioned in this ticket.

**Comment by tfennelly** [comment]
Sorry guys ... I was out in the garden for the afternoon.

@kzantow thanks a bunch for tracking that down and sorry for missing it. I can apply that fix easily and push it.

**Comment by kzantow** [comment]
@tfennelly you're welcome :) This was the only blocking issue for me. There are a number of other things mentioned in this PR that are predominantly not related to the changes here but rather the functionality as it existed, which I think can be addressed in follow-on PRs. Either way, I'm :+1: on the overall changes and haven't found anything else... I could nitpick some other things, but it would only serve to muddy up this PR further.

**Comment by gusreiber** [comment]
:bee:

**Comment by kzantow** [comment]
:+1:  :)

**Pull request merged**